_posts/2023-07-14-types-and-programming-languages.markdown

```diff
@@ -0,0 +1,610 @@
+---
+layout: post
+title:  "How to use Dafny to prove type safety"
+date:   2023-07-14 18:00:00 +0100
+author: Mikael Mayer
+---
+<script type="text/javascript" async src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
+<link rel="stylesheet" href="/blog/assets/css/types-and-programming-languages.css">
+<img class="clickable" id="img-intro" src="/blog/assets/images/type-and-programming-languages/introimage.png" alt="A type checker on the term If(True, 0, 1)" style="display:block;margin-left:auto;margin-right:auto;width:500px;max-width:95%;"/>
+There are many ways even well tested programs can go wrong (see my [previous blog post]({{ site.baseurl }}{% post_url 2023-04-19-making-verification-compelling-visual-verification-feedback-for-dafny %}) on how Dafny helps).
+Static typing often solves a class of these programming issues entirely by preventing unintended usage of a value.
+
+This blog post takes a deep dive... take a deep breath... on how to write _terms of a programming language_, and both a _type-checker_ and an _evaluator_ on such terms, such that the following _soundness property_ holds:
+
+> 1. [Progress] If a type checker accepts a term, then the evaluator will _not get stuck_ on that term.
+> 2. [Preservation] If a term has a type T, then the evaluator will also return a term of type T
+
+We will illustrate this using the infrastructure of the following Blockly workspace.
+If you build a full term and pass it to "Type check", if the type checker succeeds, the block is surrounded with green, otherwise with red.
+If surrounded with green, then attaching the term to "Evaluate" and clicking on "Evaluate" will always do something, except if the term is a final value. "Evaluate" is also surrounded in red if its term does not type check.
```

Comment:
"“Evaluate” is also surrounded ... " -> "“Evaluate” is also surrounded ... ". (not sure if the red is in addition to the green, then please ignore this)